### PR TITLE
added point collector

### PR DESCRIPTION
**Pull request by lgostling**

### CODE REVIEW

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 1]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
             }
         }
 
+        [FunctionName("PointCollector")]
+        public Task PointCollector([QueueTrigger("point")] string queueItem, ExecutionContext executionContext, ILogger logger, int dequeueCount)
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:656]: Let's call this pointcollector. I know the rest of the GH queue don't have the collector suffix, but that would make things more consistent with ADO.

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 2]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+                CollectorIdentity = this.configManager.GetCollectorIdentity(),
+                FunctionStartDate = functionStartDate,
+                SessionId = sessionId,
+                InvocationId = executionContext.InvocationId.ToString(),
+                DequeueCount = dequeueCount,
+            };
+
+            ITelemetryClient telemetryClient = new GitHubApplicationInsightsTelemetryClient(this.telemetryClient, context, logger);
+            bool success = false;
+            string outputPaths = String.Empty;
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:684]: Nit: string instead of String.

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 3]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+            Dictionary<string, string> sessionEndProperties = new Dictionary<string, string>();
+            StatsTracker statsTracker = null;
+            string identifier = "Point";
+            try
+            {
+                ICache<PointCollectorTableEntity> pointCache = new AzureTableCache<PointCollectorTableEntity>(telemetryClient, "point");
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:690]: Depending on the purpose of this cache, I would suggest calling it differently.

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 4]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+                await pointCache.InitializeAsync().ConfigureAwait(false);
+                Repository repository = pointCollectorInput.getRepository();
+                ICache<RateLimitTableEntity> rateLimiterCache = new AzureTableCache<RateLimitTableEntity>(telemetryClient, "ratelimiter");
+                await rateLimiterCache.InitializeAsync().ConfigureAwait(false);
+                IRateLimiter rateLimiter = new GitHubRateLimiter(this.configManager.UsesGitHubAuth(context.CollectorType) ? repository.OrganizationLogin : "*", rateLimiterCache, this.httpClient, telemetryClient, maxUsageBeforeDelayStarts: 95.0, this.apiDomain, throwOnRateLimit: true);
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:695]: Let's not give 95% of the usage to point collector, at least until all web hook processing is off webhook processor. Maybe start with 75-85%?

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 5]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+                await requestsCache.InitializeAsync().ConfigureAwait(false);
+                GitHubHttpClient httpClient = new GitHubHttpClient(this.httpClient, rateLimiter, requestsCache, telemetryClient);
+                IAuthentication authentication = this.configManager.GetAuthentication(CollectorType.Point, httpClient, repository.OrganizationLogin, this.apiDomain);
+
+                StorageManager storageManager;
+                List<IRecordWriter> recordWriters;
+                FunctionContextWriter<FunctionContext> contextWriter = new FunctionContextWriter<FunctionContext>();
+                using (storageManager = this.configManager.GetStorageManager(context.CollectorType, telemetryClient))
+                {
+                    recordWriters = storageManager.InitializeRecordWriters(identifier, context, contextWriter, this.adlsClient.AdlsClient);
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:706]: Hmmmm, at some point, let's revisit how we are using AdlsClient to initiate the record writers. I think this is the case in ADO collectors as well, but considering that we don't do directly ingestion to ADLS anymore (within the collectors), we should make this simpler. Can you leave a comment here so that we don't forget?

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 6]
```diff
@@ -652,6 +652,97 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+                CloudQueue trafficCloudQueue = await AzureHelpers.GetStorageQueueAsync("point").ConfigureAwait(false);
+                TimeSpan? initialVisibilityDelay = exception.getHiddenTime();
+                TimeSpan? timeToLive = null;
+                await trafficCloudQueue.AddMessageAsync(new CloudQueueMessage(queueItem), timeToLive, initialVisibilityDelay, new QueueRequestOptions(), new OperationContext()).ConfigureAwait(false);
+                telemetryClient.TrackException(exception, "RateLimiterRequeue");
+            }
+            catch (Exception exception)
+            {
+                telemetryClient.TrackException(exception);
+                throw exception;
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:737]: Alan kept fixing this for me, but using throw; instead of throw exception is the better approach. This way you don't generate another level of indirection in the stack trace :) See, Alan, I keep learning...

File: GitHub.Collectors/Cache/PointCollectorTableEntity.cs [hunk 7]
```diff
@@ -0,0 +1,34 @@
+﻿using Microsoft.CloudMine.Core.Collectors.Cache;
+using Microsoft.CloudMine.GitHub.Collectors.Model;
+
+namespace Microsoft.CloudMine.GitHub.Collectors.Cache
+{
+    public class PointCollectorTableEntity : TableEntityWithContext
```
Comment (kivancmuslu) [inline GitHub.Collectors/Cache/PointCollectorTableEntity.cs:6]: Not sure how this cache works, could you please explain? I have a fear that this might not be doing what you are looking for.

Is this attempting to cache a particular request / record so that if we are already going to process it, we would not again? If this is the case, would not keeping track of the exact URL to call (instead of org. repo, record type) easier? Also, when do we remove stuff? After all, you want to re-collect some URLs over time in case something has changed. (This might come later in the PR)

File: GitHub.Collectors/PointCollectorInput.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+namespace Microsoft.CloudMine.GitHub.Collectors
+{
+    [Serializable]
+    public class PointCollectorInput
+    {
+        public string Url { get; set; }
+        public virtual string RecordType { get; set; }
+        public virtual string ApiName { get; set; }
+        public JObject Context { get; set; }
```
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:14]: Can we make Context similar to what we have in ADO point collector, a dictionary of <string, JToken>?

File: GitHub.Collectors/PointCollectorInput.cs [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+using Newtonsoft.Json.Linq;
+using System;
+
+namespace Microsoft.CloudMine.GitHub.Collectors
+{
+    [Serializable]
+    public class PointCollectorInput
+    {
+        public string Url { get; set; }
+        public virtual string RecordType { get; set; }
```
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:12]: Do we have any instantiations of PointCollectorInput? If not, why have virtual stuff?

File: GitHub.Collectors/PointCollectorInput.cs [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        public virtual string ApiName { get; set; }
+        public JObject Context { get; set; }
+        public bool IgnoreCache { get; set; } = false;
+        public string[] IgnoreCacheForApis { get; set; } = new string[0];
+        // Optional fields to overwrite default authentication
+        public string IdentityEnvironmentVariable { get; set; } = "Identity";
+        public string PersonalAccessTokenEnvironmentVariable { get; set; } = "PersonalAccessToken";
+
+        public string getOrganizationLogin()
```
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:21]: getOrganizationLogin => GetOrganizationLogin

File: GitHub.Collectors/PointCollectorInput.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+        {
+            JToken organizationLoginToken = this.Context.SelectToken("$.OrganizationLogin");
+            if (organizationLoginToken == null)
+            {
+                throw new FatalTerminalException("Invalid request: point collector request must contain the following JSON attribute: $.Request.Context.OrganizationLogin");
+            }
+            return organizationLoginToken.Value<string>();
+        }
+
+        public Repository getRepository()
```
Comment (kivancmuslu) [inline GitHub.Collectors/PointCollectorInput.cs:31]: getRepository => GetRepository.

File: GitHub.Collectors/Processor/PointProcessor.cs [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                RecordType = input.RecordType,
+                ApiName = input.ApiName,
+                GetInitialUrl = metadata => input.Url,
+                AdditionalMetadata = additionalMetadata,
+            };
+
+            foreach (IRecordWriter recordWriter in this.recordWriters)
+            {
+                await recordWriter.NewOutputAsync(input.RecordType).ConfigureAwait(false);
+            }
```
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:63]: Do we really need  a new output just to change the output name wrt record type? I think it is fine for this to stay as Point_<blah>, what do you think?

File: GitHub.Collectors/Processor/PointProcessor.cs [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            string apiName = collectionNode.ApiName;
+            if (!input.IgnoreCache && !input.IgnoreCacheForApis.Contains(apiName))
+            {
+                // PointCollectorTableEntity tableEntity = await this.cache.RetrieveAsync(new PointCollectorTableEntity(input));
+                // TODO Logic for skipping if in cache goes here
+                bool apiCollected = false;
+                if (apiCollected)
+                {
+                    // TODO track event in telemetry
+                    return;
```
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:78]: Are you planning to implement these before merging, or do you plan to bring those in a separate PR? Regardless, as I mentioned above, I don't think this cache is working as you hope for. 

If I recall our discussion correctly, we want to have a cache because we would like to ensure that we are not doing point collection on the same URL over and over, especially over a short amount of time. Here is the concrete problem:
* microsoft organization does updates to 1000 PRs.
* GH fires 1000 events where organization URL is: github.com/microsoft
* We add 1000 messages to the point collector queue with the same request.

I think the idea is that you don't even want to add these messages at the beginning. You can do this only by keeping a cache that tells you when you have collected a particular URL latest (this to be updated / inserted by the point collector) and then not redirecting the collection to the point collector if this URL was collected recently (let's say within the last 5 minutes). This check and skip needs to be baked into Webhook processor.

In other words, point collector would be a producer for this cache whereas the webhook processor would be a consumer.

If you had something else in mind, please let me know. I might be mis-remembering.

**Review by kivancmuslu — COMMENTED** [review]
Good skeleton, missing stuff, but I am sure you are aware of this.

I really like the idea to wrap point collector into the abstract collector and treat this as a collector with a single web request. I think that simplifies a lot already.

I think the major feedback is on how cache is created and utilized. Please let me know if you have any questions.

**Comment by kivancmuslu** [comment]
@AllainPL please take a look to this when you get a chance to. Thanks!

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 14]
```diff
@@ -49,56 +49,55 @@ public class DefaultCollector : ICollector
...
+                ResponseType = responseType
             };
-            foreach (IRecordWriter recordWriter in this.RecordWriters)
-            {
-                await recordWriter.WriteRecordAsync(record, context).ConfigureAwait(false);
-            }
+
+            string queueItem = JsonConvert.SerializeObject(input);
+            CloudQueue trafficCloudQueue = await AzureHelpers.GetStorageQueueAsync("pointcollector").ConfigureAwait(false);
+            await trafficCloudQueue.AddMessageAsync(new CloudQueueMessage(queueItem), null, null, new QueueRequestOptions(), new OperationContext()).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (kivancmuslu) [reply GitHub.Collectors/Model/DefaultCollector.cs:100]: ADO collectors are definitely doing this some way, if not AzureHelpers.

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 15]
```diff
@@ -49,56 +49,55 @@ public class DefaultCollector : ICollector
...
             }
         }
 
-        private async Task ProcessUrlAsync(string requestUrl, string recordType, List<HttpResponseSignature> allowlistedResponses)
+        private async Task OffloadToPointCollector(string url, string recordType, string apiName, Repository repository, string responseType = "Array")
         {
-            HttpResponseMessage response = await this.HttpClient.GetConditionalViaETagAsync(requestUrl, recordType, this.Authentication, allowlistedResponses).ConfigureAwait(false);
-            if (response.StatusCode == HttpStatusCode.NotModified)
+            ICache<PointCollectorTableEntity> pointCache = new AzureTableCache<PointCollectorTableEntity>(this.TelemetryClient, "point");
+            await pointCache.InitializeAsync().ConfigureAwait(false);
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:71]: Similar to other caches, I would initialize this cache at higher-level e.g., functions class and then pass to Webhook processor and then DefaultCollector. This way, you don't keep initializing it (even if that is optimized for that).

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 16]
```diff
@@ -49,56 +49,55 @@ public class DefaultCollector : ICollector
...
-                    UserNotFoundResponse,
-                    ResourceNotAccessibleByIntegrationResponse,
-                };
-                await this.ProcessUrlAsync(senderUrl, UserInstanceRecordType, allowlistedResponses).ConfigureAwait(false);
+                await this.OffloadToPointCollector(senderUrl, DataContract.UserInstanceRecordType, DataContract.UsersApiName, repository, "Object").ConfigureAwait(false);
             }
         }
 
-        private async Task ProcessUrlAsync(string requestUrl, string recordType, List<HttpResponseSignature> allowlistedResponses)
+        private async Task OffloadToPointCollector(string url, string recordType, string apiName, Repository repository, string responseType = "Array")
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:68]: >string responseType = "Array" [](http://example.com/codeflow?start=121&length=29)

Any reason for using "Array" as default but then using "Object" above for both. If there is no "good" default, then, leave it empty so that every called must decide what to pass.

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 17]
```diff
@@ -49,56 +49,55 @@ public class DefaultCollector : ICollector
...
                 string senderUrl = senderUrlToken.Value<string>();
-                // There are cases where we receive this payload for removed team members, members no longer exist in GitHub. Therefore, the following call can fail with 404 not found.
-                // When using a GitHub app, the same request fails with 403: "Resource not accessible by integration" response.
-                List<HttpResponseSignature> allowlistedResponses = new List<HttpResponseSignature>()
-                {
-                    UserNotFoundResponse,
-                    ResourceNotAccessibleByIntegrationResponse,
-                };
-                await this.ProcessUrlAsync(senderUrl, UserInstanceRecordType, allowlistedResponses).ConfigureAwait(false);
+                await this.OffloadToPointCollector(senderUrl, DataContract.UserInstanceRecordType, DataContract.UsersApiName, repository, "Object").ConfigureAwait(false);
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:64]: >"Object" [](http://example.com/codeflow?start=138&length=8)

Can you please name this parameter e.g.,
```
responseType: "Object"
```

Otherwise the reviewer needs to go to the method eveytime to remember what it is used for.

File: GitHub.Collectors/Cache/PointCollectorTableEntity.cs [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        public string Url { get; set; }
+        public PointCollectorTableEntity(string url)
+        {
+            this.Url = url;
+            string urlHash = HashUtility.ComputeSha256(this.Url);
+            this.PartitionKey = urlHash;
```
Comment (kivancmuslu) [inline GitHub.Collectors/Cache/PointCollectorTableEntity.cs:13]: >this.PartitionKey = urlHash; [](http://example.com/codeflow?start=12&length=28)

I am assuming we do this mainly because Url has some characters that is not permitted to be a partition key? I wonder if this will cost us read-ability and debug-ability.

Can you potentially experiemnt with Url escaping to see whether it makes the URL safe enough to use as partition key? I am just thinking about making the table user-friendlly to explore in the future. Only if this won't take much time.

File: GitHub.Collectors/Model/DataContract.cs [hunk 19]
```diff
@@ -5,6 +5,8 @@ namespace Microsoft.CloudMine.GitHub.Collectors.Model
 {
     public class DataContract
     {
+        public const string OrganizationInstanceRecordType = "GitHub.OrgIstance";
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DataContract.cs:8]: >OrgIstance [](http://example.com/codeflow?start=69&length=10)

Can you cross-check with Sascha to see how painful it would be to fix this e.g., typo here (Instance instead of Istance)?

File: GitHub.Collectors/Processor/PointProcessor.cs [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        public async Task ProcessAsync(PointCollectorInput input)
+        {
+            Dictionary<string, JToken> additionalMetadata = new Dictionary<string, JToken>();
+            foreach (KeyValuePair<string, JToken> contextElement in input.Context)
+            {
+                additionalMetadata.Add(contextElement.Key, contextElement.Value);
+            }
+
+            Type responseType = typeof(JArray);
+            if(input.ResponseType == "Object")
```
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:48]: >input.ResponseType == "Object" [](http://example.com/codeflow?start=15&length=30)

Nit: prefer to use .Equals(..) method of string instead of ==

File: GitHub.Collectors/Processor/PointProcessor.cs [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+                RecordType = input.RecordType,
+                ApiName = input.ApiName,
+                GetInitialUrl = metadata => input.Url,
+                AdditionalMetadata = additionalMetadata,
+                ResponseType = responseType
+            };
+
+            await this.collector.ProcessAsync(collectionNode).ConfigureAwait(false);
+            PointCollectorTableEntity collectionRecord = new PointCollectorTableEntity(input.Url);
+            await this.cache.CacheAsync(collectionRecord).ConfigureAwait(false);
```
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:64]: Can you double check (code and/or end-to-end exeuciton) for the semantics of CacheAsync when the record already exists in the cache? Does it do a forced update? If that is the case, I think we are good, but good to check.

Comment (lgostling) [reply GitHub.Collectors/Processor/PointProcessor.cs:64]: Yes, I have tested that when the table entry already exists this performs an update.

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 22]
```diff
@@ -22,13 +25,11 @@ public class DefaultCollector : ICollector
...
         public static readonly HttpResponseSignature ResourceNotAccessibleByIntegrationResponse = new HttpResponseSignature(HttpStatusCode.Forbidden, "Resource not accessible by integration");
 
-        public const string OrganizationInstanceRecordType = "GitHub.OrgIstance";
-        public const string UserInstanceRecordType = "GitHub.UserInstance";
-
         protected FunctionContext FunctionContext { get; private set; }
         protected GitHubHttpClient HttpClient { get; private set; }
         protected List<IRecordWriter> RecordWriters { get; private set; }
         protected ICache<RepositoryItemTableEntity> Cache { get; private set; }
+        protected ICache<PointCollectorTableEntity> PointCache { get; private set; }
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/DefaultCollector.cs:32]: >private set; [](http://example.com/codeflow?start=70&length=12)

Nit: if these are only set in the constructor, you can remove "private set;" parts.

File: GitHub.Collectors/Model/DataContract.cs (hunk 19 above)
Comment (kivancmuslu) [reply GitHub.Collectors/Model/DataContract.cs:8]: Discussed with Sascha, we are not doing this yet. Please add a comment that at some point this needs to be fixed.

File: GitHub.Collectors/Model/CollectorFactory.cs [hunk 23]
```diff
@@ -22,13 +22,14 @@ public class CollectorFactory
                                        GitHubHttpClient httpClient,
                                        List<IRecordWriter> recordWriters,
                                        ICache<RepositoryItemTableEntity> cache,
+                                       ICache<PointCollectorTableEntity> pointCache,
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/CollectorFactory.cs:25]: >pointCache [](http://example.com/codeflow?start=73&length=10)

Maybe call this "pointcollectorcache"?

**Comment by kivancmuslu** [comment]
>             HttpResponseMessage response = await this.HttpClient.GetAsync(url, this.Authentication, DataContract.CommitInstanceApiName, allowlistedResponses).ConfigureAwait(false);

Are we planning to do this later on? Totally fine to leave it to a separate PR, just asking. Also, while doing this, can you please also clean-up passing / using HttpClient in Webhook processors and Default / Commit collectors? They should no longer need it (I think).

---
Refers to: GitHub.Collectors/Model/PushCollector.cs:109 in 880409e. [](commit_id = 880409ed13b3fedd0fdad4826396d83bf1da4f84, deletion_comment = False)

File: GitHub.Collectors/Processor/PointProcessor.cs [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+using Microsoft.CloudMine.GitHub.Collectors.Collector;
+using Microsoft.CloudMine.GitHub.Collectors.Web;
+using Newtonsoft.Json.Linq;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Microsoft.CloudMine.GitHub.Collectors.Processor
+{
+    public class PointProcessor
```
Comment (kivancmuslu) [inline GitHub.Collectors/Processor/PointProcessor.cs:16]: >PointProcessor [](http://example.com/codeflow?start=17&length=14)

Can we call this PointCollector instead of processor?

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 25]
```diff
@@ -652,6 +655,98 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
             }
         }
 
+        [FunctionName("PointCollector")]
+        public Task PointCollector([QueueTrigger("pointcollector")] string queueItem, ExecutionContext executionContext, ILogger logger, int dequeueCount)
+        {
+            return this.ExecutePointCollectorAsync(queueItem, executionContext, logger, dequeueCount);
+        }
```
Comment (kivancmuslu) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:662]: Can you please also add a "dri" variant for this? As in ADO collectors.

**Comment by kivancmuslu** [comment]
This looks good on my end, please ensure that you get either Alan's or Sascha's (or both) feedback approval as well. In case I missed anything. Thanks!

---
In reply to: [751061162](https://github.com/microsoft/CEDAR.GitHub.Collector/pull/96#issuecomment-751061162) [](http://example.com/codeflow?ancestors=751061162,751061162)

**Review by kivancmuslu — APPROVED** [review]
:shipit:

File: GitHub.Collectors.Functions/GitHubFunctions.cs [hunk 26]
```diff
@@ -652,6 +655,104 @@ public async Task Traffic([QueueTrigger("traffic")] string queueItem, ExecutionC
...
+            StatsTracker statsTracker = null;
+            string identifier = "Point";
+            try
+            {
+                ICache<PointCollectorTableEntity> pointCollectorCache = new AzureTableCache<PointCollectorTableEntity>(telemetryClient, "point");
+                await pointCollectorCache.InitializeAsync().ConfigureAwait(false);
+                Repository repository = pointCollectorInput.Repository;
+                ICache<RateLimitTableEntity> rateLimiterCache = new AzureTableCache<RateLimitTableEntity>(telemetryClient, "ratelimiter");
+                await rateLimiterCache.InitializeAsync().ConfigureAwait(false);
+                IRateLimiter rateLimiter = new GitHubRateLimiter(this.configManager.UsesGitHubAuth(context.CollectorType) ? repository.OrganizationLogin : "*", rateLimiterCache, this.httpClient, telemetryClient, maxUsageBeforeDelayStarts: 80.0, this.apiDomain, throwOnRateLimit: true);
```
Comment (AllainPL) [inline GitHub.Collectors.Functions/GitHubFunctions.cs:704]: Is there a reason for this arbitrary number `80.0`?

File: GitHub.Collectors/Collector/PointCollector.cs [hunk 27]
```diff
@@ -0,0 +1,101 @@
...
+            this.cache = cache;
+            this.telemetryClient = telemetryClient;
+            this.apiDomain = apiDomain;
+        }
+
+        public async Task ProcessAsync(PointCollectorInput input)
+        {
+            Dictionary<string, JToken> additionalMetadata = new Dictionary<string, JToken>();
+            additionalMetadata.Add("OrganizationId", input.Repository.OrganizationId);
+            additionalMetadata.Add("OrganizationLogin", input.Repository.OrganizationLogin);
```
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:46]: NIT:
```suggestion
            Dictionary<string, JToken> additionalMetadata = new Dictionary<string, JToken>
            {
                { "OrganizationId", input.Repository.OrganizationId },
                { "OrganizationLogin", input.Repository.OrganizationLogin }
            };
```

File: GitHub.Collectors/Collector/PointCollector.cs [hunk 28]
```diff
@@ -0,0 +1,101 @@
...
+            PointCollectorTableEntity tableEntity = new PointCollectorTableEntity(input.Url);
+            tableEntity = await pointCollectorCache.RetrieveAsync(tableEntity).ConfigureAwait(false);
+
+            if (tableEntity != null && DateTimeOffset.UtcNow < tableEntity.Timestamp.AddMinutes(5))
+            {
+                // has been collected in last 5 minuets, skip collection
+                return;
+            }
+
+            //PointCollectorInput input = new PointCollectorInput()
```
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:86]: Intentionally left? If yes ass some explanation :)

File: GitHub.Collectors/Model/CollectorFactory.cs [hunk 29]
```diff
@@ -18,17 +18,15 @@ public class CollectorFactory
...
                                        ICache<RepositoryItemTableEntity> cache,
+                                       ICache<PointCollectorTableEntity> pointCollectorCache,
                                        ITelemetryClient telemetryClient,
                                        string apiDomain)
         {
             return eventType switch
             {
-                "push" => new PushCollector(functionContext, authentication, httpClient, recordWriters, cache, telemetryClient, apiDomain),
-                _ => new DefaultCollector(functionContext, authentication, httpClient, recordWriters, cache, telemetryClient),
+                "push" => new PushCollector(functionContext, cache, pointCollectorCache, telemetryClient, apiDomain),
```
Comment (AllainPL) [inline GitHub.Collectors/Model/CollectorFactory.cs:28]: 🤟 I just passionately love this construction

File: GitHub.Collectors/Model/DefaultCollector.cs [hunk 30]
```diff
@@ -22,82 +26,53 @@ public class DefaultCollector : ICollector
...
-            // ToDo: kivancm: 08/09/2021, due to higher rate of 'secondary' rate limiting 429s from GitHub.com, disable the following to reduce the overall HTTP request against GitHub.com.
-            //JToken organizationUrlToken = jsonObject.SelectToken($"$.organization.url");
-            //if (organizationUrlToken != null)
-            //{
-            //    string organizationUrl = organizationUrlToken.Value<string>();
-            //    await this.ProcessUrlAsync(organizationUrl, OrganizationInstanceRecordType, allowlistedResponses: new List<HttpResponseSignature>()).ConfigureAwait(false);
-            //}
+            repository = new Repository(repository.OrganizationId, 0, repository.OrganizationLogin, string.Empty);
+
+            JToken organizationUrlToken = jsonObject.SelectToken($"$.organization.url");
```
Comment (AllainPL) [inline GitHub.Collectors/Model/DefaultCollector.cs:40]: string interpolation not needed

File: GitHub.Collectors/Collector/PointCollector.cs [hunk 31]
```diff
@@ -0,0 +1,101 @@
...
+            additionalMetadata.Add("OrganizationLogin", input.Repository.OrganizationLogin);
+
+            if (input.Repository.RepositoryId != 0)
+            {
+                additionalMetadata.Add("RepositoryId", input.Repository.RepositoryId);
+                additionalMetadata.Add("RepositoryName", input.Repository.RepositoryName);
+            }
+
+            Type responseType = typeof(JArray);
+            if( input.ResponseType.Equals("Object") )
```
Comment (AllainPL) [inline GitHub.Collectors/Collector/PointCollector.cs:55]: nit, spaces :)

**Review by AllainPL — COMMENTED** [review]
I would also review the `using` statements just at a quick glance it looks like some might be not used.

Btw we will need to sit down one day and create some automatic checks and rules + configure IDEs for everyone, as all the spaces using directives and even the Dictionary initialization should be autofixable/autodiscoverable

File: GitHub.Collectors/Model/PushCollector.cs [hunk 32]
```diff
@@ -29,17 +26,21 @@ public class PushCollector : DefaultCollector
         public const string NotFoundMessage = "Not Found";
 
         private readonly string apiDomain;
+        private FunctionContext FunctionContext;
```
Comment (kivancmuslu) [inline GitHub.Collectors/Model/PushCollector.cs:29]: Naming convention: FunctionContext => functionContext

Comment (kivancmuslu) [reply GitHub.Collectors/Model/PushCollector.cs:29]: Make readonly as well?

File: GitHub.Collectors.Functions/GitHubFunctions.cs (hunk 26 above)
Comment (AllainPL) [reply GitHub.Collectors.Functions/GitHubFunctions.cs:704]: @lgostling I'm still interested in this :P
